### ISSUE TITLE
Custom attributes

### ISSUE THREAD

**Issue by Seldom-SE** [djeedai/bevy_hanabi#262]
**Describe the solution you'd like**
A way to configure an effect to use custom attributes with given types

**Describe _why_ you want that solution. Is this related to a problem?**
I have an effect with a property for the color of the particles, which I update over time in a system. It's a base color that's blended with another color over time for each particle. Currently, when I change the property, it changes the color of all of the existing particles, which looks jarring. Instead, I want it to only set the color of new particles. I can't just store the property in `Attribute::HDR_COLOR` on `init`, because the color property is blended with another color over time. So, I would like to store the property in a custom attribute for each particle when it's initialized.

**Describe alternatives you've considered**
I don't know an alternative, though perhaps there is one.

**Additional context**
My effect:
```rust
let writer = ExprWriter::new();

// ...

let age = SetAttributeModifier::new(Attribute::AGE, writer.lit(0.).expr());
let lifetime = SetAttributeModifier::new(
    Attribute::LIFETIME,
    writer.lit(0.25).uniform(writer.lit(0.5)).expr(),
);

let lerp = ((writer.attr(Attribute::AGE) / writer.attr(Attribute::LIFETIME)).log2()
    * writer.lit(3.))
.exp2();
let color = SetAttributeModifier::new(
    Attribute::HDR_COLOR,
    ((writer.prop("color") * (writer.lit(1.) - lerp.clone())
        + writer.lit(Vec4::new(1., 0.5, 0., 0.)) * lerp)
        * writer.lit(4.))
    .expr(),
);
// ...

effects.add(
    EffectAsset::new(1024, Spawner::rate(50.0.into()), writer.finish())
        .with_name("mining")
        .with_property("color", Vec4::ZERO.into())
        // ...
        .init(age)
        .init(lifetime)
        // ...
        .update(color)
        // ...
        ,
)
```

Thanks!

**Comment by djeedai** [comment]
@Seldom-SE I'm not sure I understand sorry.

> Currently, when I change the property, it changes the color of all of the existing particles

Ok, that's by design since properties are per effect, not per particle. If you want a per-particle value, as you correctly pointed you need an attribute. Which leads me to...

> I can't just store the property in Attribute::HDR_COLOR on init, because the color property is blended with another color over time

Why is that? Is it because the attribute is used as the final color in the fragment shader, and therefore cannot _also_ be used for storing another color?

If so, indeed you need another storage attribute. I've added recently some spare attribute for users to store what they want, although only of type `f32`. But it's pretty trivial to add more for other types like `vec4<f32>` for HDR color.

Interestingly, I thought you could maybe add a `SetAttributeModifier` to the fragment shader to set the particle color just before rendering but without actually storing that value in the particle buffer. But it turns out that `SetAttributeModifier` cannot be used at all in the render shaders. I think that would potentially be a better option, because the only thing you want to pay the cost of storage is a per-particle initial color, not a final blend color that's only used for actual rendering and computed from other quantities

As a temporary workaround though, if you don't need HDR everywhere (either don't need to store the initial particle color in HDR, or don't need the final output and render target in HDR), then maybe you could use both `Attribute::COLOR` and `Attribute::HDR_COLOR`? There's some `pack` operator now in the Expressions API which allows converting between the two (although not in any released version yet).

**Comment by Seldom-SE** [comment]
> Is it because the attribute is used as the final color in the fragment shader, and therefore cannot also be used for storing another color?

Yes

> maybe you could use both `Attribute::COLOR` and `Attribute::HDR_COLOR`?

Ooh, this might work for me if I switch to the `main` branch then, thanks!

**Comment by djeedai** [comment]
I've added all variants of float vectors as custom attributes. I know it's not the best fix, but for now it should give some room until there's a more robust solution. Feel free to reopen though if you think this is not enough.

**Comment by djeedai** [comment]
Just released 0.9.0 with both changes @Seldom-SE, you shouldn't need `main` anymore.

**Comment by Seldom-SE** [comment]
Thanks!
